POSIX compliance and new features

j8r opened the pull request:
I've added new improvements to your initial work:
- POSIX compliance accross shells
- Update tldr to its newest version
- Clear all cache files
- Download all cache files (for offline use)
- Improve readability with colors and tabs
- New one line command installation
- Use wget if curl is not present

I allowed myself to add myself on the license and the credits for my work. If you find this inappropriate, you can anyway remove my name but I will feel sad.

The one line command installation need to be tested when this PR will be merged.

raylee commented:
Wow! You absolutely will get credit for your work, though I will have to read through it and test it first. Given my current schedule that may not happen until Saturday. I hope you don't mind the delay.

In the meantime, thank you for the PR!

j8r committed:
POSIX compliance and new features

- POSIX compliance accross shells
- Update tldr to its newest version
- Clear all cache files
- Download all cache files (for offline use)
- Improve readability with colors and tabs
- New one line command installation
- Use wget if curl is not present

j8r committed:
Show global tldr command

j8r commented:
No problem, thank you too for the fast reply! This will let me the time also to test more on my side since I've pushed to GitHub.

j8r committed:
Various little fixes

j8r committed:
Improve docs / Make the script portable

j8r commented:
@raylee Do you plan do review this soon?

j8r committed:
change URLs to working j8r's awaiting merging

tpo commented:
@raylee - I have read through j8r's commits in his fork [1] until commit 5edba24 [2]) and I think his commits are OK.

In general I hate doing `wget http://some.url/some.shell.script.sh | sh` which j8r added in his `README.md/documentation`, but I think that's up to your taste.

Other than that I think j8r's patches are fine and I appreciate his effort to make tldr run on other POSIX shells out there - I imagine there are BSD people that will appreciate that. 

I did not an "in depth" review neither of your code nor of j8r's patches, just quickly read through everything from start to end.

So here you have my 2 cents. Best greets and thanks to both of you @raylee and @j8r,
*t

[1] https://github.com/j8r/tldr
[2] https://github.com/j8r/tldr/commit/5edba24cf78f6095ce28aaa86e23777627253c96

j8r commented:
@tpo Thanks for your review. :)

For the autoinstall script, I understand that it can be a security risk, but it is very practical for new comers. Therefore I've also added manual instructions. Overall everything is up to the end user.

I quite disagree with you for `~/bin` because tldr is not only an executable but also a directory with subfolders with a cache. For me a `~./tldr` is a good directory, with inside `tldr.sh` and the `cache |` folder.

Overall, I haven't encountered major issues but I've observed that when the `cache` folder is not present and we run tldr (that will create the cache), nothing is printed. We need to run it again.

raylee commented on README.md:
```diff
@@ -1,35 +1,55 @@
-	mkdir -p ~/bin
-	curl -o ~/bin/tldr https://raw.githubusercontent.com/raylee/tldr/master/tldr
-	chmod +x ~/bin/tldr
+`wget -qO- https://raw.githubusercontent.com/DFabric/DPlatform-ShellCore/master/install.sh | sh`
```
I'm uncomfortable with this even being an option. I don't want to encourage this sort of installation method.

j8r replied:
I really would like tldr to be the most accessible to the wider range of people.
I agree to not encourage this practice, but please keep (not highlighted if you want) this feature for people (I admit like me) who just wants to have the things done as efficiently as possible.
This is not as evil (at least here), specially for such an open source script that isn't ran privileged.

raylee commented on README.md:
```diff
@@ -1,35 +1,55 @@
 
 The client caches a copy of all pages and the index locally under
-~/.config/tldr. By default, the cached copies will expire in 14 days.
+~/.tldr. By default, the cached copies will expire in 14 days.
```
I'd rather keep this .config/tldr. This is the freedesktop.org standard for where config directories should be stored. If there were a .rc file that we wanted to store (for user configuration), that would be appropriate in ~/ (ie ~/.tldr as a file, not a directory).

tpo replied:
I was actually reflecting about this as well. Not only is ~/.config a freedesktop.org standard (which I appreciate), but there's also ~/.cache from the same standard.

I was thinking about whether the cached tldr *.md files should possibly go to ~/.cache/tldr. I think they actually should. I consider stuff in ~/.cache "ad hoc reproducible" stuff - here in the sense that it can be re-fetched at any time, if internet is available.

Also ~/.cache is something that I would not want to backup, which is the case here for *.md pages that can be trivially and painlessly re-downloaded from the tldr git repo.

So I'd argue that the cached *.md pages should actually go into ~/.cache/tldr instead.

j8r replied:
I agree we can do this it was my first idea, but I've finally chosen this way to have a true portable script, that we can bring on any directory and execute it, with a `cache` folder that will be created.

raylee replied:
~/.cache/tldr/ makes more sense, let's go with that.

j8r replied:
Ok good. I've an idea, we can add a `--portable|-p` argument with no cache, or in `/tmp`, it will solves the problem!

j8r replied:
For me we should change the argument `-p|--platform`  to `-o|--os` like on other tldr implementations like [here](https://github.com/tldr-pages/tldr-node-client/blob/master/bin/tldr#L22). We could also add `--linux`,  `--osx` and `--sunos`.

raylee commented on install.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/sh
+
+
+[ ! -d ~/.tldr ] && printf "tldr seems to have already been previously installed.\nDelete the \'$HOME/.tldr\' directory first before relaunching this script\n" && exit 1 || mkdir -p ~/.tldr/cache
```
One of the goals I tried to maintain with the script is that it should "just work" at all times. So the script should clean up after itself and maintain its own caches. This is another benefit of using ~/.config/tldr/, as it is more clearly a directory under the script's control.

j8r replied:
Indeed it's not very needed, I've deleted this in a further commit (commit [bb3a00658f74f49c3271621e467193b91c2a9241](https://github.com/j8r/tldr/commit/bb3a00658f74f49c3271621e467193b91c2a9241))

raylee commented on install.sh:
```diff
@@ -0,0 +1,27 @@
+
+[ ! -d ~/.tldr ] && printf "tldr seems to have already been previously installed.\nDelete the \'$HOME/.tldr\' directory first before relaunching this script\n" && exit 1 || mkdir -p ~/.tldr/cache
+
+download() {
```
I like the idea of a download command that covers all the cases, nice work!

raylee commented on install.sh:
```diff
@@ -0,0 +1,27 @@
+[ ! -d ~/.tldr ] && printf "tldr seems to have already been previously installed.\nDelete the \'$HOME/.tldr\' directory first before relaunching this script\n" && exit 1 || mkdir -p ~/.tldr/cache
+
+download() {
+  hash curl 2>/dev/null && curl -sf -o "$1" "$2" || { hash wget 2>/dev/null && wget -qO "$1" "$2"; } || echo "Fail to download $2" && exit 1
```
But I'm confused as to why you have `hash` in front of the commands. I don't think it's necessary, is it?

j8r replied:
I think it's better to check if `curl` or `wget` are present to avoid the user to install the one or the other package. Often `wget` is already present by default, but if `curl` is present, better to use it because it's faster (for me at least) and I also prefer it, but this is only my personal opinion.

raylee commented on install.sh:
```diff
@@ -0,0 +1,27 @@
+download ~/.tldr/cache/index.json http://tldr-pages.github.io/assets/index.json
+
+# Add alias to rc files
+add_alias() {
```
I'm uncomfortable with this. Another goal for this script was such that it could be included in distros directly with no more dependencies than the core files. Part of that means being a good citizen with the standards they ask, which includes not automatically modifying the user's profile files. So I think this belongs either in the manual instructions or as an 'install' target on on the command line which does all the magic.

As an aside, these aliases are unnecessary in the original instructions, as tldr is a command directly in their path:

`curl -o ~/bin/tldr https://raw.githubusercontent.com/raylee/tldr/master/tldr`

j8r replied:
Agree, specially because `tldr` is a single standalone file. By the way I thinks it's would be better on `/usr/bin` when tldr will have it's package, but actually even more on `/usr/local/bin`.

raylee commented on tldr.sh:
```diff
@@ -1,58 +1,56 @@
+# Improvements added by Julien Reichardt https://github.com/j8r
 # a Sunday afternoon's project, I'm sure there's room for improvement. PRs welcome!
-
-set -uf -o pipefail
```
Why are you removing this?

j8r replied:
Because this cause some shells like `dash` (`set: Illegal option -o pipefail`) . It appears to be not POSIX compliant, and I haven't noticed problems after deleting, so I remove it.

raylee commented on tldr.sh:
```diff
@@ -1,58 +1,56 @@
     index="$configdir/index.json"
     cache_days=14
-    force_update=''
+    force_update=
```
Did you test this change?

j8r replied:
Yes, this reset the same way the variable (from 'yes').

raylee commented on tldr.sh:
```diff
@@ -72,57 +72,60 @@ init_term_cmds() {
 
 heading() {
     local line="$*"
-    echo "$bold$red${line:2}$reset"
```
So, when I originally wrote the coloring and formatting I was matching the official tldr output. Maybe I got it wrong? Same comment with the other color and style changes below.

j8r replied:
You have interpreted the markdown strictly "as is". I've followed the [actual official tldr image with its colors](https://raw.githubusercontent.com/tldr-pages/tldr/master/screenshot.png), that interpret this character changes to colors, that I think is a more human readable way to transmit informations.
I've taken the liberty to make some little changes, specially at the top for the command title and its description. I think we can better differentiate them from the main contents.
Overall this colors changes are opinionated and the final thought is up to each one.

raylee commented on tldr.sh:
```diff
@@ -72,57 +72,60 @@ init_term_cmds() {
 code() {
     local line="$*"
     # I'm sure there's a better way to strip the first and last characters.
-    line="${line:1}"
```
Can you provide some commentary to explain why you're changing this? It works correctly already, so not sure why you're changing this?

j8r replied:
Because of `dash` again. I'm not sure if this substitution is POSIX or not (or what verison) — `bash --posix` and `ksh` works but a `shellcheck` shows that it isn't POSIX.
Anyway, I remove the first and the latest "`" character. We can replace "${line%\'`'}" by "${line%?}" (just tested).

raylee commented on tldr.sh:
```diff
@@ -72,57 +72,60 @@ init_term_cmds() {
-    line=${line//\{\{/$italic}
-    line=${line//\}\}/$eitalic}
+    # Check if the shell support buil-in substitutions
+    (${a//}) 2>/dev/null && line=${line//\{\{/$reset$white} || { hash sed 2>/dev/null && line="$(printf "$line" | sed "s/{{/$reset$white/g" )"; }
```
Nicely done. Same comment about `hash` though.

j8r replied:
Thanks, but i'm not specially proud of this hack. I've found that `(${a//})` return an error on `dash`, but not on other modern shells. Here we can remove the `hash sed 2>/dev/null` check because this utility is virtually present in all UN*X machine (even more than `wget` that is a "standalone" package)

raylee commented on tldr.sh:
```diff
@@ -72,57 +72,60 @@ init_term_cmds() {
     # last line even if it doesn't have a newline at the end
-    while IFS= read -r line || [[ -n "$line" ]]; do
-        start=${line:0:1}           # get the first character
+    while read line || [ -n "$line" ]; do
```
So this change remove the IFS= which prevents while from stripping whitespace. Rationale?

j8r replied:
Now I'm also asking myself this questions. It's not really needed... It's nevertheless a good practice?

raylee commented on tldr.sh:
```diff
@@ -156,49 +159,62 @@ path_for_cmd() {
-    }
-    # if the curl failed for some reason, keep cat from whinging
-    cat "$cached" 2>/dev/null
+    recent "$cached" || { mkdir -p $(dirname $cached); download "$cached" "$base_url/$p"; }
```
Please split across multiple lines as per the original.

raylee commented on tldr.sh:
```diff
@@ -207,21 +223,32 @@ exit 0
+            ;;
         -l|--list)
-            >&2 echo -e "Known tldr pages: \n"
-            tr '{' '\n' < "$configdir/index.json" | cut -d '"' -f4 | column
```
Fair enough, I guess the user can pipe it to `column` on their own.

tpo replied:
even after 20 years of unix `column` is a command that I'm still only half aware of.

raylee replied:
@tpo Heh, I find it one of those 'obscure but occasionally useful' commands, along with `join` and a few others.

tpo replied:
```
$ tldr join
tldr page for command join not found
```

damn!

raylee commented on tldr.sh:
```diff
@@ -233,14 +260,18 @@ do
     exit 1
 fi
 
-display_tldr <<< "$tldr"
```
For my own education, is <<< not supported in all POSIX shells?

j8r replied:
You are right :)

raylee commented on install.sh:
```diff
@@ -0,0 +1,30 @@
+#!/bin/sh
```
Another unstated goal was that everything would be a single file, which keeps the installation easy. I could see adding all the alias stuff to an install target in tldr itself, (tldr -i).

raylee reviewed (changes requested):
Lots of really good work! I like a lot of the directions you are heading, but I do have a bunch of detailed comments on the PR.

I think many of them are simply due to my lack of shell scripting experience. So please keep that in mind when you read my comments, I think I'll just need a bit of education as to why you made some of the changes as they aren't immediately apparent to me.

Thank you for giving such care and love to this little weekender project!

j8r commented:
I've made various changes according to your requests, thank you @raylee to have build this project from scratch, on which I have based my work. Thanks to you and @tpo  to take time to review my code!

I've kept `install.sh` and `README.md`unmodified, modify them as your convenience.
I've also added a `-p|--portable` argument for those who want to execute `tldr` on the go.

j8r committed:
Various code improvements

j8r committed:
Move alias additions to the tldr file

j8r commented:
@raylee and @tpo, I've finalized the changes you have requested. Tell me if the PR is good to merge now, thank you again!

raylee commented:
Good morning! I'll be reviewing these changes in the next couple days, thank you for both your patience and your continued efforts.

j8r committed:
Fix bash substitutions + improve docs

j8r commented:
@raylee Have you reviewed the changes?

raylee commented:
I've given it a first pass and it looks good. The last thing I'd like to think about is making the upgrade clean for users, in particular removing the old cached pages if they exist. Thoughts?

j8r commented:
Also if we use `upgrade` when`tldr` is in `/usr/local/bin`, we need to change `chmod +x` to `chmod +wx`.
By the way I agree, we can do this like you said, and also add a `remove` feature if you are OK.

raylee commented:
You've been patient enough on this, let's leave the upgrade issues for a separate PR. So for now, I have no requests for changes.

raylee commented:
So I just want to check with you on your intent. Here's a screenshot of the rendering before the changes in this PR:

![image](https://cloud.githubusercontent.com/assets/715888/20723409/49f72914-b61f-11e6-9a74-f31ece17cb0b.png)

And here it is afterward:

![image](https://cloud.githubusercontent.com/assets/715888/20723400/39ad19a6-b61f-11e6-85c6-4a2de641c0d8.png)

(This is with the same source md file in each case, naturally.) Does this match your intent? Some of the color choices are difficult with a bright background.

There is another open PR that I'm going to look at which makes the color choices configurable via environment variables. If it's okay with you, I'm going to cherry-pick the changes from your fork, and allow you to override the colors via the other PR.

j8r commented:
We should create a bright/dark terminal mode, because we will have problems like this.
Also I've closely matched the "*official*" colors (including bright ones).
Anyway I think it's simpler to merge this, because the modifications brought by this PR are quite numerous compared to the color patch.
Then we can work with @ewels to solve this color issue, and *cherry-pick* his fixes.
But the final decision is yours.

j8r commented:
I'm just thinking, most macOS users (like you) have bright terminal and most of Linux users (like me) have dark ones.
So why not use this to set right defaults to the colors, of course in addition of the @ewels 's PR?

raylee commented:
Cherry-picking isn't a big deal. There's no need for a themes, just stick with all color choices of the same contrast/saturation.

Anyway, there's a reason why it's usually one change per PR. Let's aim for that in the future. Thanks!

j8r commented:
There are pros and cons of both configurations.
Nice output but not universal, or universal dull output?
IMHO we should open an issue on this to let the users say what they really would like.
This would be interesting to see how other tldr clients handle this.

ewels commented:
I think boring but customisable colours is going to be the best way to go. If someone tries this script and can't read the output on the first go, they're probably not going to hang around trying to figure out how to change them. However, making it possible (and preferably simple) to change the colours is definitely a worthwhile feature.

If the readme has a bunch of "themes" with screenshots and cut+paste env variables, then that should be fine I think? Or even named presets such as `dark` and `light` etc. Could save this with the cache instead of in environment variables? Sorry, just thinking out loud here :)

tpo commented:
@ewels - let's keep in mind, that tldr is also and defacto "just a little shell script" where opening it in a text editor, scrolling through the script, finding the place where the colors are set and changing them right there does not amount to rocket science, espacially if the place where the color setting happens would be nicely commented...

ewels commented:
Sorry @tpo, but I disagree. For me the entire spirit of _tldr_ is that it's beginner friendly. Editing a bash script that's hundreds of lines long so that you can read the output doesn't really seem to fit that ethos. Having a handful of cut and paste `export VAR` statements in the readme isn't really adding much complexity to the code and keeps things simple for those who are new to a command line environment.
